module/zcommon/zfs_fletcher_aarch64_neon.c

```diff
@@ -60,141 +60,185 @@ static void
     const void *buf, uint64_t size)
 {
-	const uint64_t *ip = buf;
-	const uint64_t *ipend = (uint64_t *)((uint8_t *)ip + size);
+	const uint32_t *ip = buf;
+	const uint32_t *ipend = ip + (size / sizeof (uint32_t));
 #if defined(_KERNEL)
-register unsigned char ZERO asm("v0") __attribute__((vector_size(16)));
-register unsigned char ACC0 asm("v1") __attribute__((vector_size(16)));
-register unsigned char ACC1 asm("v2") __attribute__((vector_size(16)));
-register unsigned char ACC2 asm("v3") __attribute__((vector_size(16)));
-register unsigned char ACC3 asm("v4") __attribute__((vector_size(16)));
-register unsigned char TMP1 asm("v5") __attribute__((vector_size(16)));
-register unsigned char TMP2 asm("v6") __attribute__((vector_size(16)));
-register unsigned char SRC asm("v7") __attribute__((vector_size(16)));
+	register unsigned char AA2 asm("v1") __attribute__((vector_size(16)));
+	register unsigned char A3A4 asm("v2") __attribute__((vector_size(16)));
+	register unsigned char BB2 asm("v3") __attribute__((vector_size(16)));
+	register unsigned char B3B4 asm("v4") __attribute__((vector_size(16)));
+	register unsigned char CC2 asm("v5") __attribute__((vector_size(16)));
+	register unsigned char C3C4 asm("v6") __attribute__((vector_size(16)));
+	register unsigned char DD2 asm("v7") __attribute__((vector_size(16)));
+	register unsigned char D3D4 asm("v8") __attribute__((vector_size(16)));
+	register unsigned char SRC asm("v9") __attribute__((vector_size(16)));
+	register unsigned char ZERO asm("v10") __attribute__((vector_size(16)));
+	register unsigned char TMP1 asm("v11") __attribute__((vector_size(16)));
+	register unsigned char TMP2 asm("v12") __attribute__((vector_size(16)));
 #else
-unsigned char ZERO __attribute__((vector_size(16)));
-unsigned char ACC0 __attribute__((vector_size(16)));
-unsigned char ACC1 __attribute__((vector_size(16)));
-unsigned char ACC2 __attribute__((vector_size(16)));
-unsigned char ACC3 __attribute__((vector_size(16)));
-unsigned char TMP1 __attribute__((vector_size(16)));
-unsigned char TMP2 __attribute__((vector_size(16)));
-unsigned char SRC __attribute__((vector_size(16)));
+	unsigned char AA2 __attribute__((vector_size(16)));
+	unsigned char A3A4 __attribute__((vector_size(16)));
+	unsigned char BB2 __attribute__((vector_size(16)));
+	unsigned char B3B4 __attribute__((vector_size(16)));
+	unsigned char CC2 __attribute__((vector_size(16)));
+	unsigned char C3C4 __attribute__((vector_size(16)));
+	unsigned char DD2 __attribute__((vector_size(16)));
+	unsigned char D3D4 __attribute__((vector_size(16)));
+	unsigned char SRC __attribute__((vector_size(16)));
+	unsigned char ZERO __attribute__((vector_size(16)));
+	unsigned char TMP1 __attribute__((vector_size(16)));
+	unsigned char TMP2 __attribute__((vector_size(16)));
 #endif
 
 	kfpu_begin();
 
 	NEON_INIT_LOOP();
 
-	for (; ip < ipend; ip += 2) {
-		NEON_MAIN_LOOP(NEON_DONT_REVERSE);
+	for (; ip < ipend; ip += 4) {
+		NEON_MAIN_LOOP_LOAD(NEON_DO_NOT_REVERSE);
```

Comment:
It might be helpful if we had the assembly/disassembly from this loop to study. For example, Clang generates for a version of this code that I will probably put into a PR soon:

```
.LBB0_2:                                // =>This Inner Loop Header: Depth=1
        ldr     q16, [x1], #16
        uaddw   v4.2d, v4.2d, v16.2s
        uaddw2  v0.2d, v0.2d, v16.4s
        add     v1.2d, v0.2d, v1.2d
        add     v5.2d, v4.2d, v5.2d
        add     v6.2d, v5.2d, v6.2d
        add     v2.2d, v1.2d, v2.2d
        add     v3.2d, v2.2d, v3.2d
        add     v7.2d, v6.2d, v7.2d
        cmp     x1, x8
        b.lo    .LBB0_2
```

I am not an expert on ARM assembly, but this looks as tight as possible.

Reply:
Here is a sandbox you can use to see the disassembly generated by clang:

https://gcc.godbolt.org/z/aP9ff3jPh

I realize that the x86-64 “version” of clang was used, but the beauty of LLVM is that one binary can support multiple architectures by setting the target. Since that site does not have a clang explicitly for aarch64/arm, I set a target as a compiler flag instead.

You can grab the byteswap version from #14234 and put it there to see what is generated for the byteswap case.